### PR TITLE
Introducing aca-py clients structure, with first Java client (aka controller) implementation

### PR DESCRIPTION
**Pull request by etschelp**
We work a lot with aca-py, but as most of our projects are written in Java I created a small library that lets you interact with aca-py directly from Java e.g.

`acaPy.connectionsReceiveInvitation(invitation, alias);`

The first idea was to make a pull request in the controllers repo, but the client is less about writing examples, instead its tightly coupled to aca-py's rest interface and it's supposed to evolve with it (versioning and all). So I propose a structure similar to whats in indy-sdk/wrappers where the sdk also houses all language specific implementations as well. Together with the structure I also want to introduce the first client implementation for Java. The term _client_ can be interchanged with _controller_ as well.

The Java client is not feature complete, and not a 1:1 mapping, it's more about simplification and getting the most important use cases up and running.

### CODE REVIEW

**Comment by swcurran** [comment]
@sklump , @andrewwhitehead -- can you please take a look at this and the overall approach?  Is this a good way to enable a multi-language stack?  Perhaps we can discuss this at the next ACA-Pug -- next Wednesday (11AM Pacific).

File: clients/java/src/main/java/org/hyperledger/aries/api/connection/Connection.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package org.hyperledger.aries.api.connection;
+
+import org.apache.commons.lang3.StringUtils;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.Accessors;
+
+@Data @NoArgsConstructor @Accessors(chain = true)
+public class Connection {
```
Comment (sklump) [inline clients/java/src/main/java/org/hyperledger/aries/api/connection/Connection.java:15]: ConnectionRecord?

File: clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import com.google.gson.JsonObject;
+import com.google.gson.annotations.SerializedName;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+
+@Data @NoArgsConstructor @AllArgsConstructor
+public class PresentProofPresentation {
```
[the remark this answers is not in the document]
Comment (etschelp) [reply clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java:22]: Makes sense, I will rename this class

File: clients/java/src/main/java/org/hyperledger/aries/api/connection/Connection.java (hunk 1 above)
Comment (etschelp) [reply clients/java/src/main/java/org/hyperledger/aries/api/connection/Connection.java:15]: I will rename this one as well.

File: clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (etschelp) [reply clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java:22]: Yes, issue-credential methods are mostly missing. My plan was to add aca-py interactions step by step. But if missing methods block the pull request I will prioritize them.

[the remark this answers is not in the document]
Comment (sklump) [reply clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java:22]: If the idea is only to use it as a verifier, then issue-credential becomes unnecessary. Overall I think the approach is fine.

[the remark this answers is not in the document]
Comment (etschelp) [reply clients/java/src/main/java/org/hyperledger/aries/api/proof/PresentProofPresentation.java:22]: Right now yes. But finally the client is supposed to support issue-credential as well, starting with the automated flows (--auto-xxx flags)

**Comment by chaeso** [comment]
Great job! One thing I noticed is that there's no way to create schema with your library

**Comment by etschelp** [comment]
@chaeso I quickly added the /schemas endpoint as part of the latest commit.

**Comment by chaeso** [comment]
I think `Invitation` object used for connectionsReceiveInvitation() method is not correct.

Example for parameter used for `receive-invitation` API must be as follows. But `Invitation` object does not compatible with it.
`{
  "@id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
  "recipientKeys": [
    "H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
  ],
  "imageUrl": "http://192.168.56.101/img/logo.jpg",
  "label": "Bob",
  "routingKeys": [
    "H3C2AVvLMv6gmMNam3uVAjZpfkcJCwDwnZn6z3wXmqPV"
  ],
  "did": "WgWxqztrNooG92RXvxSTWv",
  "serviceEndpoint": "http://192.168.56.101:8020"
}`

**Comment by etschelp** [comment]
Good point. I extended the class to contain all fields that are part of the request. Most of the time one does not need to send all fields in a request for it to be accepted by aca-py, and in this case the only mandatory field is the did (to make a connection), so the request model was compressed down to that. What I will do is to use annotations wherever possible to make this fact more clear.

**Comment by swcurran** [comment]
@etschelp -- can you join us on the ACA-Pug User Group call this coming Wednesday (Aug. 5 11AM Pacific -- e.g. super late for you :-( ) to discuss this?  The big issue is should this be its on repo or should it be within the ACA-Py repo.

The concern is keeping it up to date and how we maintain synchronization.

If not at that time, I'd like to be able to get you and the team together on another day -- at a better time for you :-).

**Comment by etschelp** [comment]
As discussed at the last [users group meeting](https://wiki.hyperledger.org/display/ARIES/2020-08-05+Aries+Cloud+Agent+-+Python+Users+Group+Community+Meeting), I moved the clients folder into a new repository available here: https://github.com/boschresearch/aries-acapy-clients, so that we can make it ready to be moved to hyperledger.

**Pull request closed without merging**